orchagent/pbhorch.cpp

```diff
@@ -577,57 +665,227 @@ bool PbhOrch::updatePbhRule(const PbhRule &rule)
+        attr.value.aclfield.mask.u8 = rule.ipv6_next_header.mask;
+
+        if (!pbhRule->validateAddMatch(attr))
+        {
+            SWSS_LOG_ERROR("Failed to configure PBH rule(%s) match: IPV6_NEXT_HEADER", rule.key.c_str());
+            return false;
+        }
+    }
+
+    if (rule.l4_dst_port.is_set)
+    {
+        sai_attribute_t attr;
+
+        attr.id = SAI_ACL_ENTRY_ATTR_FIELD_L4_DST_PORT;
+        attr.value.aclfield.enable = true;
+        attr.value.aclfield.data.u16 = rule.l4_dst_port.value;
+        attr.value.aclfield.mask.u16 = rule.l4_dst_port.mask;
+
+        if (!pbhRule->validateAddMatch(attr))
+        {
+            SWSS_LOG_ERROR("Failed to configure PBH rule(%s) match: L4_DST_PORT", rule.key.c_str());
+            return false;
+        }
+    }
+
+    if (rule.inner_ether_type.is_set)
+    {
+        sai_attribute_t attr;
+
+        attr.id = SAI_ACL_ENTRY_ATTR_FIELD_INNER_ETHER_TYPE;
+        attr.value.aclfield.enable = true;
+        attr.value.aclfield.data.u16 = rule.inner_ether_type.value;
+        attr.value.aclfield.mask.u16 = rule.inner_ether_type.mask;
+
+        if (!pbhRule->validateAddMatch(attr))
+        {
+            SWSS_LOG_ERROR("Failed to configure PBH rule(%s) match: INNER_ETHER_TYPE", rule.key.c_str());
+            return false;
+        }
     }
 
-    if (!this->aclOrch->updateAclRule(rule.table, rule.name, rule.flow_counter.value))
+    if (rule.hash.is_set && rule.packet_action.is_set)
+    {
+        PbhHash hObj;
+
+        if (this->pbhHlpr.getPbhHash(hObj, rule.hash.value))
+        {
+            sai_attribute_t attr;
+
+            attr.id = rule.packet_action.value;
+            attr.value.aclaction.enable = true;
+            attr.value.aclaction.parameter.oid = hObj.getOid();
+
+            if (!pbhRule->validateAddAction(attr))
+            {
+                SWSS_LOG_ERROR("Failed to configure PBH rule(%s) action", rule.key.c_str());
+                return false;
+            }
+        }
+    }
+
+    if (!pbhRule->validate())
+    {
+        SWSS_LOG_ERROR("Failed to validate PBH rule(%s)", rule.key.c_str());
+        return false;
+    }
+
+    // Mellanox W/A
+    if (this->pbhCap.getAsicVendor() == PbhAsicVendor::MELLANOX)
+    {
+        const auto &hMeta = rule.hash.meta;
+        const auto &paMeta = rule.packet_action.meta;
+
+        auto cond1 = std::find(uFields.cbegin(), uFields.cend(), hMeta.name) != uFields.cend();
+        auto cond2 = std::find(uFields.cbegin(), uFields.cend(), paMeta.name) != uFields.cend();
+
+        if (cond1 || cond2)
+        {
+            auto pbhRulePtr = dynamic_cast<AclRulePbh*>(this->aclOrch->getAclRule(rule.table, rule.name));
+
+            if (pbhRulePtr == nullptr)
+            {
+                SWSS_LOG_ERROR("Failed to update PBH rule(%s) in SAI: invalid object type", rule.key.c_str());
+                return false;
+            }
+
+            if (!pbhRulePtr->disableAction())
```

Comment:
So if I understand the work around correctly, we will disable the acl  rule and then update the acl rule, which will as part of the update re-enable the acl with the newly defined rule?

Reply:
PBH rule configuration defines `packet_action` and `hash` attributes as:
1. SAI_ACL_ENTRY_ATTR_ACTION_SET_LAG_HASH_ID
2. SAI_ACL_ENTRY_ATTR_ACTION_SET_ECMP_HASH_ID

Which means that two PBH fields refer to a single separate action from SAI point of view.

**Example:**
```json
"PBH_RULE": {
    "pbh_table|nvgre": {
        "priority": "2",
            "ether_type": "0x0800",
            "ip_protocol": "0x2f",
            "gre_key": "0x2500/0xffffff00",
            "inner_ether_type": "0x86dd",
            "hash": "inner_v6_hash",
            "packet_action": "SET_ECMP_HASH",
            "flow_counter": "DISABLED"
    }
}
```

In order to implement a W/A for Mellanox platform, we need to disable both actions first, before moving to update (`packet_action` or `hash`) operation via ACL orch:

```cpp
bool AclRulePbh::disableAction()
{
    const auto &cit1 = m_actions.find(SAI_ACL_ENTRY_ATTR_ACTION_SET_ECMP_HASH_ID);
    if (cit1 != m_actions.cend())
    {
        const auto &attr1 = cit1->second.getSaiAttr();
        if (attr1.value.aclaction.enable)
        {
            sai_attribute_t attr;

            attr.id = attr1.id;
            attr.value.aclaction.enable = false;
            attr.value.aclaction.parameter.oid = SAI_NULL_OBJECT_ID;

            if (!setAttribute(attr))
            {
                return false;
            }

            m_actions.erase(cit1);
        }
    }

    const auto &cit2 = m_actions.find(SAI_ACL_ENTRY_ATTR_ACTION_SET_LAG_HASH_ID);
    if (cit2 != m_actions.cend())
    {
        const auto &attr2 = cit2->second.getSaiAttr();
        if (attr2.value.aclaction.enable)
        {
            sai_attribute_t attr;

            attr.id = attr2.id;
            attr.value.aclaction.enable = false;
            attr.value.aclaction.parameter.oid = SAI_NULL_OBJECT_ID;

            if (!setAttribute(attr))
            {
                return false;
            }

            m_actions.erase(cit2);
        }
    }

    return true;
}
```